contracts/src/token/erc20/extensions/erc4626.rs

```diff
@@ -0,0 +1,528 @@
+            }));
+        }
+        let assets = self.preview_mint(shares);
+        self._deposit(msg::sender(), receiver, assets, shares)?;
+        Ok(assets)
+    }
+
+    fn max_withdraw(&self, owner: Address) -> U256 {
+        self._convert_to_assets(self._asset.balance_of(owner), Rounding::Floor)
+    }
+
+    fn preview_withdraw(&self, assets: U256) -> U256 {
+        self._convert_to_shares(assets, Rounding::Ceil)
+    }
+
+    fn withdraw(
+        &mut self,
+        assets: U256,
+        receiver: Address,
+        owner: Address,
+    ) -> Result<U256, Error> {
+        let max_assets = self.max_withdraw(owner);
+        if assets > max_assets {
+            return Err(Error::ExceededMaxWithdraw(
+                ERC4626ExceededMaxWithdraw { owner, assets, max: max_assets },
+            ));
+        }
+
+        let shares = self.preview_redeem(assets);
+        self._withdraw(msg::sender(), receiver, owner, assets, shares)?;
+        Ok(shares)
+    }
+
+    fn max_redeem(&self, owner: Address) -> U256 {
+        self._asset.balance_of(owner)
+    }
+
+    fn preview_redeem(&self, shares: U256) -> U256 {
+        self._convert_to_assets(shares, Rounding::Ceil)
+    }
+
+    fn redeem(
+        &mut self,
+        shares: U256,
+        receiver: Address,
+        owner: Address,
+    ) -> Result<U256, Error> {
+        let max_shares = self.max_redeem(owner);
+        if shares > max_shares {
+            return Err(Error::ExceededMaxRedeem(ERC4626ExceededMaxRedeem {
+                owner,
+                shares,
+                max: max_shares,
+            }));
+        }
+
+        let assets = self.preview_redeem(shares);
+        self._withdraw(msg::sender(), receiver, owner, assets, shares)?;
+        Ok(assets)
+    }
+}
+
+impl Erc4626 {
+    fn _convert_to_shares(&self, assets: U256, rounding: Rounding) -> U256 {
+        //assets._mul_div_(  self.total_assets() + 10 **
+        // self._decimals_offset(), self.total_assets() + 1, rounding)
+        U256::ZERO
+    }
+
+    fn _convert_to_assets(&self, shares: U256, rounding: Rounding) -> U256 {
+        //shares.mul_div(x, y, dominator, rounding)
+        U256::ZERO
+    }
+
+    fn _deposit(
+        &mut self,
+        caller: Address,
+        receiver: Address,
+        assets: U256,
+        shares: U256,
+    ) -> Result<(), Error> {
+        // If _asset is ERC-777, `transferFrom` can trigger a reentrancy BEFORE
+        // the transfer happens through the `tokensToSend` hook. On the
+        // other hand, the `tokenReceived` hook, that is triggered after the
+        // transfer, calls the vault, which is assumed not malicious.
+        //
+        // Conclusion: we need to do the transfer before we mint so that any
+        // reentrancy would happen before the assets are transferred and
+        // before the shares are minted, which is a valid state.
+        // slither-disable-next-line reentrancy-no-eth
+
+        self._asset._mint(receiver, shares)?;
```

Comment:
Please take great care to call ERC-20 functions on _the appropriate token_!

In this case, you should be [calling `_mint` on ERC4626 itself](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC4626.sol#L248), and _not_ on the asset!

Double check all of the other functions.

Reply:
i understand